examples/end_to_end.py

```diff
@@ -1,39 +1,109 @@
+import numpy as np
 import torch
 import torch.nn as nn
-import numpy as np
-from torch_radon import Radon
-
-batch_size = 8
-n_angles = 64
-image_size = 128
-channels = 4
+import torch.nn.functional as F
+from torch.utils.data import Dataset
+from torch_radon import FanBeam, Volume2D
 
-device = torch.device('cuda')
-criterion = nn.L1Loss()
 
-# Instantiate a model for the sinogram and one for the image
-sino_model = nn.Conv2d(1, channels, 5, padding=2).to(device)
-image_model = nn.Conv2d(channels, 1, 3, padding=1).to(device)
+class TestDataset(Dataset):
+    def __init__(self):
+        super().__init__()
+        img = np.load('phantom.npy')
+        
+        imgs = [
+            img,
+            np.fliplr(img).copy(),
+            np.rot90(img, k=1).copy(),
+            np.rot90(img, k=2).copy(),
+            np.rot90(img, k=3).copy(),
+        ]
+        self.imgs = [_ for _ in imgs] + [(1.0 - _.copy()) for _ in imgs]
+    
+    def __len__(self):
+        return len(self.imgs)
 
-# create empty images
-x = torch.FloatTensor(batch_size, 1, image_size, image_size).to(device)
+    def __getitem__(self, index):
+        img = self.imgs[index]
+        x = torch.from_numpy(img).unsqueeze(0).float()
+        return x
 
-# instantiate Radon transform
-angles = np.linspace(0, np.pi, n_angles)
-radon = Radon(image_size, angles)
 
-# forward projection
-sinogram = radon.forward(x)
+class TestNet(nn.Module):
+    def __init__(self, channels=3) -> None:
+        super().__init__()
+        self.sino_model = nn.Sequential(
+            nn.Conv2d(1, channels, 5, padding=2),
+            nn.Conv2d(channels, 1, 5, padding=2)
+        )
+        self.img_model = nn.Conv2d(2, 1, 3, padding=1).to(device)
+        self.det_count = 672
+    
+    def forward_sino(self, sino):
+        print('>>> Forwarding sino net.')
+        return self.sino_model(sino)
+    
+    def forward_img(self, img1, img2):
+        print('>>> Forwarding img net.')
+        img = torch.cat([img1, img2], dim=1)
```

Comment:
Can't we avoid non-contiguous arrays by not concatenating along dim 1 in this function?

Reply:
I tried this (and adding `.contiguous` for the other tensors) and it didn't seem to work. So this is most likely triggered in the backprojection process, not the external tensor definitions.

Reply:
Interesting.

Reply:
This concatenation causes con-contiguity because if I remove it, the contiguity error disappears.

```python
        img_pred = self.img_model(img_sino)
```